firebase-auth/src/androidMain/kotlin/dev/gitlive/firebase/auth/auth.kt

```diff
@@ -22,71 +24,125 @@ actual class FirebaseAuth internal constructor(val android: com.google.firebase.
-            }
-        }
-        android.addAuthStateListener(listener)
-        awaitClose { android.removeAuthStateListener(listener) }
-    }
     actual suspend fun signInWithCredential(authCredential: AuthCredential) =
         AuthResult(android.signInWithCredential(authCredential.android).await())
 
     actual suspend fun signOut() = android.signOut()
-}
 
-actual class AuthCredential(val android: com.google.firebase.auth.AuthCredential)
+    actual suspend fun updateCurrentUser(user: FirebaseUser) = android.updateCurrentUser(user.android).await().run { Unit }
+    actual suspend fun verifyPasswordResetCode(code: String): String = android.verifyPasswordResetCode(code).await()
+}
 
 actual class AuthResult internal constructor(val android: com.google.firebase.auth.AuthResult) {
     actual val user: FirebaseUser?
         get() = android.user?.let { FirebaseUser(it) }
 }
 
-actual class FirebaseUser internal constructor(val android: com.google.firebase.auth.FirebaseUser) {
-    actual val uid: String
-        get() = android.uid
-    actual val displayName: String?
-        get() = android.displayName
-    actual val email: String?
-        get() = android.email
-    actual val phoneNumber: String?
-        get() = android.phoneNumber
-    actual val isAnonymous: Boolean
-        get() = android.isAnonymous
-    actual suspend fun delete() = android.delete().await().run { Unit }
-    actual suspend fun reload() = android.reload().await().run { Unit }
-    actual suspend fun sendEmailVerification() = android.sendEmailVerification().await().run { Unit }
+actual class ActionCodeResult(val android: com.google.firebase.auth.ActionCodeResult) {
+    actual val operation: Operation
+        get() = when (android.operation) {
+            com.google.firebase.auth.ActionCodeResult.PASSWORD_RESET -> Operation.PasswordReset
+            com.google.firebase.auth.ActionCodeResult.VERIFY_EMAIL -> Operation.VerifyEmail
+            com.google.firebase.auth.ActionCodeResult.RECOVER_EMAIL -> Operation.RecoverEmail
+            com.google.firebase.auth.ActionCodeResult.ERROR -> Operation.Error
+            com.google.firebase.auth.ActionCodeResult.SIGN_IN_WITH_EMAIL_LINK -> Operation.SignInWithEmailLink
+            com.google.firebase.auth.ActionCodeResult.VERIFY_BEFORE_CHANGE_EMAIL -> Operation.VerifyBeforeChangeEmail
+            com.google.firebase.auth.ActionCodeResult.REVERT_SECOND_FACTOR_ADDITION -> Operation.RevertSecondFactorAddition
+            else -> Operation.Error
+        }
+    actual fun <T, A: ActionCodeDataType<T>> getData(type: A): T? = when (type) {
+        is ActionCodeDataType.Email -> android.info?.email
+        is ActionCodeDataType.PreviousEmail -> (android.info as? ActionCodeEmailInfo)?.previousEmail
+        is ActionCodeDataType.MultiFactor -> (android.info as? ActionCodeMultiFactorInfo)?.multiFactorInfo?.let { MultiFactorInfo(it) }
+        else -> null
+    } as? T
```

Comment:
I like the way you have implemented this! Could you take it one step further by getting rid of the type argument? you can use a reified type parameter to check what type is expected.  Also is passing the wrong type to be expected at runtime? If not (it's only a case of programmer error) the return type should not be optional instead use as! and let it throw if its the wrong type

Reply:
I simplified slightly by making the ActionCodeResult the sealed class